hash_table/hash_table.go

```diff
@@ -0,0 +1,134 @@
+package hash_table
+
+import (
+	"errors"
+)
+
+type Node struct {
+	key   int
+	value int
+	next  *Node
+}
+
+type HashTable struct {
+	arr  []*Node
+	size int
+}
+
+func NewHashTable() *HashTable {
+	return &HashTable{
+		arr: make([]*Node, 10),
+	}
+}
+
+func (h *HashTable) Insert(key int, value int) {
+	// same load factor as Java 10
+	if h.getLoadFactor() >= 0.75 {
+		h.rehash(true)
+	}
+
+	h.size++
+	hash := h.getHash(key)
+	node := h.arr[hash]
+
+	// first node in the linked list
+	if node == nil {
+		node := &Node{
+			key:   key,
+			value: value,
+		}
+		h.arr[hash] = node
+		return
+	}
+
+	// find last node in the linked list
+	for node.next != nil {
+		node = node.next
+	}
+
+	node.next = &Node{
+		key:   key,
+		value: value,
+	}
+
+}
+
+func (h *HashTable) Search(key int) (int, error) {
+	hash := h.getHash(key)
+	node := h.arr[hash]
+
+	if node == nil {
+		return -1, errors.New("Key not found")
+	}
+
+	for node.key != key {
+		node = node.next
+	}
+
+	return node.value, nil
+}
+
+func (h *HashTable) Delete(key int) (int, error) {
```

Comment:
Delete function also doesn't throw error if key can't be found.

Reply:
My intention is to allow users of this struct to know whether the operation succeeds or not.

Reply:
After calling this function, there is no value with given key in hash table. I think this function will always success even the key exists or not.